packages/jupyterlab-voila/src/preview.tsx

```diff
@@ -0,0 +1,106 @@
+import {
+  MainAreaWidget,
+  IFrame,
+  ToolbarButton,
+  ReactWidget
+} from "@jupyterlab/apputils";
+import { DocumentRegistry } from "@jupyterlab/docregistry";
+import { INotebookModel } from "@jupyterlab/notebook";
+
+import * as React from "react";
+
+export const VOILA_ICON_CLASS = "jp-MaterialIcon jp-VoilaIcon";
+
+export namespace VoilaPreview {
+  export interface IOptions extends MainAreaWidget.IOptionsOptionalContent {
+    url: string;
+    label: string;
+    context: DocumentRegistry.IContext<INotebookModel>;
+    renderOnSave: boolean;
+  }
+}
+
+export class VoilaPreview extends MainAreaWidget<IFrame> {
+  constructor(options: VoilaPreview.IOptions) {
+    super({
+      ...options,
+      content: new IFrame({ sandbox: ["allow-same-origin", "allow-scripts"] })
+    });
+
+    let { url, label, context, renderOnSave } = options;
+
+    this.content.url = url;
+    this.content.title.label = label;
+    this.content.title.icon = VOILA_ICON_CLASS;
+    this.content.id = `voila-${++Private.count}`;
+
+    const reloadButton = new ToolbarButton({
+      iconClassName: "jp-RefreshIcon",
+      onClick: this.reload,
+      tooltip: "Reload Preview"
+    });
+
+    const renderOnSaveCheckbox = ReactWidget.create(
+      <label>
+        <input
+          style={{ verticalAlign: "middle" }}
+          name="renderOnSave"
+          type="checkbox"
+          defaultChecked={renderOnSave}
+          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+            this.renderOnSave = event.target.checked;
+          }}
+        />
+        Render on Save
+      </label>
+    );
+    renderOnSaveCheckbox.addClass("jp-VoilaPreview-renderOnSave");
+
+    this.toolbar.addItem("reload", reloadButton);
+    this.toolbar.addItem("renderOnSave", renderOnSaveCheckbox);
```

Comment:
jlab/phosphorjs supports react components?

Reply:
Yes (for jlab), most of the toolbar components are actually built with React (using the `ReactWidget` utility under the hood)